force_wfmanager/left_side_pane/evaluator_model_view.py

```diff
@@ -178,35 +200,67 @@ def _create_slots_tables(self):
     def _update_slots_tables(self):
         """ Update the tables of slots when a change on the model triggers a
         change on the shape of the input/output slots """
+        self.input_slots_representation[:] = []
+        self.output_slots_representation[:] = []
+
         input_slots, output_slots = self._evaluator.slots(self.model)
 
         #: Initialize the input slots
-        self.input_slots_representation = []
-        self.model.input_slot_maps = []
-        for input_slot in input_slots:
-            self.model.input_slot_maps.append(InputSlotMap(name=''))
+        self.model.input_slot_maps = [
+            InputSlotMap(name='')
+            for _ in input_slots
+        ]
 
         #: Initialize the output slots
-        self.output_slots_representation = []
         self.model.output_slot_names = len(output_slots)*['']
 
         self._fill_slot_rows(input_slots, output_slots)
 
     def _fill_slot_rows(self, input_slots, output_slots):
         """ Fill the tables rows according to input_slots and output_slots
         needed by the evaluator and the model slot values """
+        if isinstance(self.model, BaseDataSourceModel):
+            available_variables = \
+                self.variable_names_registry.data_source_available_variables
+        else:
+            available_variables = \
+                self.variable_names_registry.kpi_calculator_available_variables
+
         self.input_slots_representation = [
-            InputSlotRow(index=index,
+            InputSlotRow(model=self.model,
+                         available_variables=available_variables,
+                         index=index,
                          name=self.model.input_slot_maps[index].name,
-                         type=input_slot.type,
-                         model=self.model)
+                         type=input_slot.type)
             for index, input_slot in enumerate(input_slots)
         ]
 
         self.output_slots_representation = [
-            OutputSlotRow(index=index,
+            OutputSlotRow(model=self.model,
+                          index=index,
                           name=self.model.output_slot_names[index],
-                          type=output_slot.type,
-                          model=self.model)
+                          type=output_slot.type)
             for index, output_slot in enumerate(output_slots)
         ]
+
+    @on_trait_change(
+        'variable_names_registry.data_source_available_variables[]')
+    def update_data_source_input_rows(self):
+        if not isinstance(self.model, BaseDataSourceModel):
+            return
+
+        available_variables = \
+            self.variable_names_registry.data_source_available_variables
+        for input_slot_row in self.input_slots_representation:
+            input_slot_row.available_variables = available_variables
```

Comment:
Ok, so you are pushing the information in. I wonder if it would be a good idea to let the input_slot_rows fetch it instead, passing the registry to them.